### PR TITLE
[macos] [metal] Fix use after free in FlutterExternalTextureMetal

### PR DESCRIPTION
**Pull request by iskakaushik**
This also adds a test for the said failure

Fixes: https://github.com/flutter/flutter/issues/78056

### CODE REVIEW

File: shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm [hunk 1]
```diff
@@ -18,6 +19,47 @@
...
+  NSDictionary* options = @{
+    // This key is required to generate SKPicture with CVPixelBufferRef in metal.
+    (NSString*)kCVPixelBufferMetalCompatibilityKey : @YES
+  };
+  CVPixelBufferRef pxbuffer = NULL;
+  CVReturn status =
+      CVPixelBufferCreate(kCFAllocatorDefault, _width, _width, kCVPixelFormatType_32BGRA,
+                          (__bridge CFDictionaryRef)options, &pxbuffer);
+
+  NSParameterAssert(status == kCVReturnSuccess && pxbuffer != NULL);
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm:55]: You aren't checking parameters of this method. So I think `NSAssert` makes more sense.

File: shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm [hunk 2]
```diff
@@ -18,6 +19,47 @@
 #include "third_party/skia/include/core/SkSamplingOptions.h"
 #include "third_party/skia/include/core/SkSurface.h"
 
+@interface TestExternalTexture : NSObject <FlutterTexture>
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm:22]: Just double checking, all TUs are ARC enabled right? Reviewing with that assumption.

File: shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm [hunk 3]
```diff
@@ -18,6 +19,47 @@
...
+}
+
+- (CVPixelBufferRef)pixelBuffer {
+  NSDictionary* options = @{
+    // This key is required to generate SKPicture with CVPixelBufferRef in metal.
+    (NSString*)kCVPixelBufferMetalCompatibilityKey : @YES
+  };
+  CVPixelBufferRef pxbuffer = NULL;
+  CVReturn status =
+      CVPixelBufferCreate(kCFAllocatorDefault, _width, _width, kCVPixelFormatType_32BGRA,
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm:52]: Are CoreVideo objects ARC enabled?

**Review by chinmaygarde — APPROVED** [review]

File: shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm (hunk 2 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm:22]: Yup, they are

File: shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm [hunk 4]
```diff
@@ -18,6 +19,47 @@
...
+                          (__bridge CFDictionaryRef)options, &pxbuffer);
+
+  NSParameterAssert(status == kCVReturnSuccess && pxbuffer != NULL);
+
+  CVPixelBufferLockBaseAddress(pxbuffer, 0);
```
[the remark this answers is not in the document]
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterEmbedderExternalTextureUnittests.mm:57]: I copy-pasted this from `FlutterScenarioTestTexture`, will fix in both areas.

**Pull request merged**